Dockerfile

```diff
@@ -64,85 +56,89 @@ RUN apt-get update && \
         /usr/share/doc \
         /usr/share/doc-base
 
-ADD hack/recurl.sh .
-ADD hack/image_arch.sh .
-RUN . ./image_arch.sh && ./recurl.sh /usr/local/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v${KUBECTL_VERSION}/bin/${IMAGE_OS}/${IMAGE_ARCH}/kubectl
+COPY hack/recurl.sh .
+RUN if [ $(uname -m) = ppc64le ] || [ $(uname -m)  = s390x ]; then \
+        ./recurl.sh docker.tgz https://download.docker.com/$(uname -s|tr '[:upper:]' '[:lower:]')/static/${DOCKER_CHANNEL}/$(uname -m) docker-18.06.3-ce.tgz; \
+    else \
+        ./recurl.sh docker.tgz https://download.docker.com/$(uname -s|tr '[:upper:]' '[:lower:]')/static/${DOCKER_CHANNEL}/$(uname -m)/docker-${DOCKER_VERSION}.tgz; \
+    fi && \
+    tar --extract --file docker.tgz --strip-components 1 --directory /usr/local/bin/ && \
+    rm docker.tgz
+
+RUN ./recurl.sh /usr/local/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v${KUBECTL_VERSION}/bin/$(uname -s|tr '[:upper:]' '[:lower:]')/$(uname -m)/kubectl
 RUN ./recurl.sh /usr/local/bin/jq https://github.com/stedolan/jq/releases/download/jq-${JQ_VERSION}/jq-linux64
 RUN rm recurl.sh
 
-COPY hack/ssh_known_hosts /etc/ssh/ssh_known_hosts
-COPY hack/nsswitch.conf /etc/nsswitch.conf
-COPY --from=builder /usr/local/bin/docker /usr/local/bin/
+COPY hack/ssh_known_hosts /etc/ssh/
+COPY hack/nsswitch.conf /etc/
+
 
 ####################################################################################################
 
 FROM node:14.0.0 as argo-ui
 
-ADD ["ui", "ui"]
-ADD ["api", "api"]
+COPY ui/package.json ui/yarn.lock ui/
 
 RUN JOBS=max yarn --cwd ui install --network-timeout 1000000
+
+COPY ui ui
+COPY api api
+
 RUN JOBS=max yarn --cwd ui build
 
 ####################################################################################################
-# Argo Build stage which performs the actual build of Argo binaries
+
+FROM builder as argoexec-build
+
+RUN --mount=type=cache,target=/root/.cache/go-build make dist/argoexec
+
 ####################################################################################################
-FROM builder as argo-build
 
-ARG IMAGE_OS=linux
+FROM builder as workflow-controller-build
 
-# Perform the build
-WORKDIR /go/src/github.com/argoproj/argo
-COPY . .
-# check we can use Git
-RUN git rev-parse HEAD
+RUN --mount=type=cache,target=/root/.cache/go-build make dist/workflow-controller
```

Comment:
this is a buildx way to do build caching, I'm not 100% sure it works, so we may want to remove later on